mlir/lib/IR/MLIRContext.cpp

```diff
@@ -188,7 +188,11 @@ class MLIRContextImpl {
 
   /// This is a sorted container of registered operations for a deterministic
   /// and efficient `getRegisteredOperations` implementation.
-  SmallVector<RegisteredOperationName, 0> sortedRegisteredOperations;
+  SmallVector<std::pair<StringRef, RegisteredOperationName>, 0>
```

Comment:
I don't follow why we need this StringRef here. The dialect name is already available from the RegisteredOperationName, isn't it?

Reply:
I don't quite understand what you mean here? I probably should checkout the PR locally and build it to figure out myself the problem.

Reply:
Essentially, as you can notice we have pairs (D1, O1), (D2, O2)... and so on (where D=DialectName and O=OperationName).

Now, while inserting I am getting the lower bound for (Di, Oi) which is the next pair to insert so it gets fit in the right place in the segment of operations for a specific dialect, i.e, (D, O1), (D, O2), (D, Oi), here all D's are the same.

For retrieval, we do it in a similar way.

If we don't use pairwise elements we wouldn't know which segment (dialect class) of the sorted elements list the operation goes to since we only have the count of operations for each dialect class (via the string map I added), but we need at least one index, either start or end.

Reply:
I don't see why you need the dialect name stored this way, there are many ways around it. The easiest would be getting the dialect name from the RegisteredOperationName for comparison in the lower bound: `getDialect().getNamespace()`. (There are other ways as well, e.g. comparing a string directly against the operation name). That would remove the need to have the StringRef here, or the `transformedOperations` field.

Reply:
Yeah, that's much better, I missed it :)